src/uchen/tboard/tboard_file.cc

```diff
@@ -1,25 +1,167 @@
+  }
+
+  std::string time_in_seconds = std::to_string(time(nullptr));
+
+  std::string filename = absl::StrFormat("%s.out.tfevents.%s.%s", log_name,
+                                         time_in_seconds, hostname);
+  std::filesystem::path file_path =
+      events_directory / std::filesystem::path(filename);
+  std::ofstream file(file_path, std::ios::binary);
+
   if (!file.is_open()) {
     return absl::InternalError("Failed to open file");
   }
   return TBoardFile(std::move(file));
 }
 
-void TBoardFile::RecordLoss(float loss) {
-  LOG(INFO) << "Recording loss: " << loss;
-  file_ << "Loss: " << loss << std::endl;
+void TBoardFile::AddScalar(const std::string& tag, double value,
+                           uint32_t step) {
+  LOG(INFO) << "Adding scalar: " << tag << ": " << value << ", Step: " << value;
+
+  tensorflow::Summary summary;
+
+  tensorflow::Summary::Value* summary_value = summary.add_value();
+  summary_value->set_tag(tag);
+  summary_value->set_simple_value(value);
+
+  AddEvent(summary, step);
+}
+
+void TBoardFile::AddImage(const std::string tag, const std::string& image,
+                          uint32_t step, uint32_t width, uint32_t height,
+                          uint32_t channel, const std::string& image_name,
+                          const std::string& image_description) {
+  auto encoded_image = EncodeImage(image);
+  if (!encoded_image.ok()) {
+    LOG(ERROR) << "Failed to Add image \"" << image_name
+               << "\": " << encoded_image.status();
+    return;
+  }
+
+  LOG(INFO) << "Recording image: " << tag;
+
+  tensorflow::SummaryMetadata metadata;
+  if (!image_name.empty()) {
+    metadata.set_display_name(image_name);
+  } else {
+    metadata.set_display_name(tag);
+  }
+  metadata.set_summary_description(image_description);
+
+  tensorflow::Summary::Image image_summary;
+  image_summary.set_width(width);
+  image_summary.set_height(height);
+  image_summary.set_colorspace(channel);
+  image_summary.set_encoded_image_string(encoded_image.value());
+
+  tensorflow::Summary summary;
+  tensorflow::Summary::Value* summary_value = summary.add_value();
+  summary_value->set_tag(tag);
+  summary_value->mutable_image()->Swap(&image_summary);
+  summary_value->mutable_metadata()->Swap(&metadata);
+
+  AddEvent(summary, step);
+}
+
+void TBoardFile::AddEvent(tensorflow::Summary& summary, uint32_t step) {
+  tensorflow::Event event;
+  event.set_wall_time(static_cast<double>(time(nullptr)));
+  event.set_step(step);
+  event.mutable_summary()->Swap(&summary);
+
+  std::string buffer;
+  event.SerializeToString(&buffer);
+  size_t buffer_length = buffer.size();
+
+  uint32_t crc_buffer_length =
+      crc32c::Crc32c((uint8_t*)&buffer_length, sizeof(size_t));
+  crc_buffer_length = TBoardFile::Mask(crc_buffer_length);
+
+  uint32_t crc_buffer = crc32c::Crc32c((uint8_t*)buffer.data(), buffer_length);
+  crc_buffer = TBoardFile::Mask(crc_buffer);
+
+  file_.write((char*)&buffer_length, sizeof(size_t));
```

Comment:
```suggestion
  file_.write(static_cast<char*>(&buffer_length), sizeof(size_t));
```

Please use appropriate C++ cast. This helps prevent UB.

Reply:
Also, why the char? Char can be signed on some platforms or even be over 1 byte long.

I do not the file format, so I apologize if the comment is wrong.